src/placeos-frontend-loader/api/remote/remote.cr

```diff
@@ -0,0 +1,150 @@
+module PlaceOS::FrontendLoader
+  class Metadata
+    private getter hash_file : HashFile = HashFile
+    private getter lock : Mutex = Mutex.new
+
+    class_getter instance : Metadata do
+      new
+    end
+
+    def with_metadata
+      lock.synchronize do
+        yield hash_file
+      end
+    end
+  end
+
+  abstract class PlaceOS::FrontendLoader::Remote
+    private alias Git = PlaceOS::Compiler::Git
+    private alias Remote = PlaceOS::FrontendLoader::Remote
+
+    getter metadata : Metadata
+
+    def initialize(@metadata : Metadata = Metadata.instance)
+    end
+
+    struct Commit
+      include JSON::Serializable
+      getter commit : String
+      getter date : String
+      getter author : String
+      getter subject : String
+
+      def initialize(@commit, @date, @author, @subject)
+      end
+    end
+
+    struct Reference
+      include JSON::Serializable
+
+      enum Type
+        GitLab
+        Github
+      end
+
+      getter repo_name : String
+      getter remote_type : Reference::Type
+      getter branch : String
+      getter tag : String | Nil
+
+      def initialize(@url : String, @branch : String? = "master", @tag : String? = nil, @hash : String? = "HEAD")
+        @repo_name = @url.split(".com/").last.rstrip("/")
+
+        if @url.includes?("github")
+          @remote_type = Type::Github
+        elsif @url.includes?("gitlab")
+          @remote_type = Type::GitLab
+        else
+          raise Exception.new("URL not supported")
+        end
```

Comment:
I know it's pedantic.... buuuuut

github.com/gitlab
gitlab.com/github


I think you should match on `URI.parse(@url).host`

Reply:
Also above you're doing `.split(".com/")`, where if you had a `URI` instance, you could obtain it via `uri.path`